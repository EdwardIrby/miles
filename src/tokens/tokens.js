/**
 * @param {...{}} objs - object containing camelCased
 * style properties or css custom properties
 * @returns {<{'--prop': 'value'}>} containing prefixed css
 *  custom properties to be inlined.
 * @example Simple
 * <Textblock style={tokens({
*    lineHeight: 1.2,
*    fontSize: rem(47.78),
*  })} />
 * @example Conditional Example
 * <Textblock style={tokens(
 *  inline ? { display: 'inline' } : { display: 'block' },
 *  {
 *    lineHeight: 1.2,
 *    fontSize: rem(47.78),
 *  },
 * )} />
 */
export const tokens = (...objs) => objs
  .filter(Boolean)
  .map(obj => Object.entries(obj))
  .reduce((acc, cur) => acc.concat(cur), [])
  .reduce((acc, [key, value]) => ({ ...acc, [`--${key}`]: value }), {});
