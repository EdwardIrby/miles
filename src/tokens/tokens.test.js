import { test, assert } from '@dxworks/rite';
import { tokens } from './tokens.js';

test('tokens()', () => {
  const expected = {
    '--width': 32,
    '--height': 24,
    '--backgroundColor': 'black',
  };
  const actual = tokens(
    {
      width: 32,
      height: 24,
      backgroundColor: 'black',
    },
  );
  assert({
    given: 'object as argument',
    should: 'return object',
    actual,
    expected,
  });
});

test('tokens() conditional test', () => {
  const checked = false;
  const disabled = true;
  const expected = {
    '--width': 32,
    '--height': 24,
    '--backgroundColor': 'grey',
  };
  const actual = tokens(
    {
      width: 32,
      height: 24,
      backgroundColor: 'black',
    },
    checked && {
      backgroundColor: 'blue',
    },
    disabled && {
      backgroundColor: 'grey',
    },
  );
  assert({
    given: 'conditional arguments',
    should: 'return object',
    actual,
    expected,
  });
});
