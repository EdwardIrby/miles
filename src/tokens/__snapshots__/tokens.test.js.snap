// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tokens util function 1`] = `
Object {
  "--backgroundColor": "black",
  "--height": 24,
  "--width": 32,
}
`;

exports[`Tokens util function conditional 1`] = `
Object {
  "--backgroundColor": "grey",
  "--height": 24,
  "--width": 32,
}
`;
